### PR TITLE
fix set_fied actions

### PR DESCRIPTION
**Pull request by hkwi**
I created ryu certificate-based test cases for this. Thanks to @gizmoguy, https://github.com/hkwi/ryutest_adhoc/blob/master/of13/gh/27_push_vid.json is the test case.

### CODE REVIEW

**Comment by gizmoguy** [comment]
Hey @hkwi great to see the test case.

I still think this PR needs to be rebased against the latest Dev_064 branch. 

We reworked a lot of the VLAN functions in PR #22 to simplify the VLAN code and remove a lot of ifs and memcpys. At the moment this PR will undo a lot of this work by removing the push and pop vlan helper methods.

Once you have rebased this PR against the latest code I am more than happy to review and test in my ZodiacFX lab for peer review.

**Comment by hkwi** [comment]
I don't know why you have trouble with this patch set, becase github system says "This branch has no conflicts with the base branch".

**Comment by gizmoguy** [comment]
Yes, it doesn't have any conflicts because it looks like git merge commit has happened at some point in your branch which allows the commit to have no conflicts with the base branch.

However if you look at the actual commit you are submitting to us as a pull request you will see it removes a lot of code we don't want to remove.

An interactive rebase against Dev_064 will let you continue with this PR and get the right code merged. However since the changes are quite significant you could look at cutting a new branch on your fork of Dev_064 and locally applying your smaller diff that adds your changes.

Let me know if you want help with doing either option.

File: ZodiacFX/src/openflow_13.c [hunk 1]
```diff
@@ -90,346 +90,317 @@ static inline uint64_t (htonll)(uint64_t n)
...
-	{
-		TRACE("Pop VLAN");
-		memmove(p_uc_data + 12, p_uc_data + 16, *packet_size - 16);
-		fields->eth_prot = *(uint16_t*)(p_uc_data + 12);
-		*packet_size -= 4;
-		*ul_size = *packet_size;
-		packet_fields_parser(p_uc_data, fields);
-        }
-}
-
```
Comment (gizmoguy) [inline ZodiacFX/src/openflow_13.c]: Shouldn't remove push_vlan_helper() and pop_vlan_helper()

Comment (hkwi) [inline ZodiacFX/src/openflow_13.c]: Why? I removed and put back inline because the final code of set_field does not reference the helper functions.

**Comment by gizmoguy** [comment]
There's something I'm not quite understanding here, so bear with me while I try to understand your changes.

So I think the only commit I'm puzzled by is 4bd2ccf2e70e3de74c0f2fcb423265b84be2ea0e. Could you explain what was broken in the set_field action that you have now fixed, and what the reason for inlining push_vlan_helper() and pop_vlan_helper() was. Because we did use them in the set_field action.

**Comment by hkwi** [comment]
From the spec:

> The use of a set-ﬁeld action assumes that the corresponding header ﬁeld exists in the packet, the eﬀect of a set-ﬁeld action on a packet that does not have the corresponding header ﬁeld is undeﬁned (switch and ﬁeld dependent). In particular, when setting the VID on a packet without a VLAN header, a switch may or may not automatically add a new VLAN tag, and the controller must explicitly use a Push VLAN header action to be compatible with all switch implementations. 

So, we can remove vlan push/pop operation from set_field, which is faster. If we want to support the operation regardless with that downside, it will also OK and I would include the code.

The broken things was:
- "Instruction set" processing order
- access with struct ofp13_instruction_actions was invalid
- struct packet_fields must be updated on packet rewrite
    - if the member was pointer, update must happen on vlan push/pop
    - if the member was immediate value, update must happen on each field update
- recalculate_ip_checksum must be done before OFPAT_OUTPUT
- recalculate_ip_checksum must not set for ARP packets

**Comment by gizmoguy** [comment]
Right I think I see what you're doing now. Since you only push or pop VLANs in one place now, you have inlined push_vlan_helper() and pop_vlan_helper().

Not sure if we want to keep those functions around for future use, they do make the code a little easier to read. I'll defer to @pzanna or @anarkiwi to make the final call.

I have tested your patch set and looks good against my Faucet test lab.

Quick question about your QinQ support, should I be able to push two VLAN tags and match two VLAN tags from OpenFlow? Or do your changes just support QinQ traffic passing through the device?

**Comment by pzanna** [comment]
At this point I think it makes sense to roll them in, I'll merge the pull request.

**Comment by hkwi** [comment]
For QinQ Answer:

VLAN tags may be pushed multiple times by specification. In the openflow match, the "outermost" one will match and the inner vlan will be transparent while outer was not popped.

Openflow spec 1.3.5

5.11 List of Actions
> The eﬀect of those actions is cumulative, if the list of actions contains two Push VLAN actions, two VLAN headers are added to the packet.

5.12 Actions
>  Set-Field actions should always be applied to the outermostpossible header (e.g. a “Set VLAN ID” action always sets the ID of the outermost VLAN tag), unless the ﬁeld type speciﬁes otherwise. 

7.2.3.8 Header Match Fields
>  If not explicitly speciﬁed in the ﬁeld description, each ﬁeld type refers to the outermost occurrence of the ﬁeld in the packet headers.

**Comment by gizmoguy** [comment]
Yeah that's what I expected to happen. I couldn't get two VLAN tags pushed however.

I made a FlowMod that had an action like:

```
actions = [
            parser.OFPActionPushVlan(0x8100),
            parser.OFPActionSetField(vlan_vid=200|ofproto_v1_3.OFPVID_PRESENT),
            parser.OFPActionPushVlan(0x88a8),
            parser.OFPActionSetField(vlan_vid=100|ofproto_v1_3.OFPVID_PRESENT),
            parser.OFPActionOutput(1), parser.OFPActionOutput(2)
        ]
```

And according to the ZodiacFX trace mode the flow matched but I don't get a packet with double tagging out the otherside and then the FlowMod mysteriously disappears from the flow table.

Am I doing something wrong?

**Comment by hkwi** [comment]
I've tested the double-push with ryu [test](https://github.com/hkwi/ryutest_adhoc/blob/master/of13/ex/00_qinq_push.json) , and the [result](https://github.com/hkwi/ZodiacFX_RyuCert/blob/master/v064_adhoc.log) was OK.

What is your ZodiacFX code hash ? May I look into ... ? The strange thing is "flow disappears". That reminds me that I removed a [strange code] (https://github.com/NorthboundNetworks/ZodiacFX/pull/30/files#diff-8e61d1caa77b081f1b1f9afe87889ce0L149) in this PR.

**Comment by gizmoguy** [comment]
Hey @hkwi just wanted to update you.

I tried 0.64 firmware just now and it works now, I must have been missing those other commits you did.

Nice work on this patch, 802.1ad working nicely now.

```
09:26:27.483760 52:54:00:ff:b2:66 > ff:ff:ff:ff:ff:ff, ethertype 802.1Q (0x8100), length 68: vlan 100, p 0, 
                     ethertype 802.1Q, vlan 200, p 0, ethertype ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 172.31.1.2 tell 
                     172.31.1.1, length 46

```

**Pull request merged**